examples/mixture/plot_gmm_init.py

```diff
@@ -0,0 +1,190 @@
+"""
+==========================
+GMM Initialization Methods
+==========================
+
+Examples of the different methods of initialization in Gaussian Mixture Models
+
+See :ref:`gmm` for more information on the estimator.
+
+Here we generate some sample data with four easy to identify clusters.The
+purpose of this example is to show the four different methods for the
+initialization parameter *init_param*.
+
+- *kmeans* - performs kmeans clustering on the dataset to produce the initial
+  centers. This will be the most resource intensive initialization but will
+  often produce intitial means that are very close to the converged gmm model.
+- *k-means++* - uses the initialization of kmeans clustering (called k-means++)
+  to use input data points as the initial centers but to pick them such that
+  they are sufficiently far apart from each other.
+- *rand_data* - uses randomly selected data points as the initial centers. This
+  is a fast method but can potentially lead to unstable results if all of the
+  selected points are close together.
+- *random* - Is the simplest method. It uses centers that are close to the mean
+  of the data. This can be the slowest initialization to converge.
+
+Orange diamonds represent the initialization centers for the gmm generated by
+the *ini_param*. The rest of the data is represented as crosses and the
+colouring represents the eventual associated classification after the gmm has
+finished.
+
+The numbers in the top right of each subplot represent the number of
+iterations taken for the GaussianMixture to converge and the time taken for the
+initialization part of the algorithm to run. The shorter initialization times
+tend to have a greater number of iterations to converge.
+
+The initialization time is the ratio of the time taken for that method versus
+the time taken for the *random* method. As you can see *random* and *rand_data*
+are on similar orders of time taken but in general *k-means++* will be slower
+and *kmeans* slower still.
+
+In this example, when initialized with *rand_data* or *random* the model takes
+more iterations to converge.
+"""
+
+
+# Author Gordon Walsh (github: g-walsh) gordon.p.walsh@gmail.com
+# data generation code from https://jakevdp.github.io/
+
+import matplotlib.pyplot as plt
+import numpy as np
+from sklearn.cluster import KMeans
+from sklearn.mixture import GaussianMixture
+from sklearn.utils.extmath import row_norms
+from sklearn.cluster.k_means_ import _k_init
```

Comment:
We really shouldn't be using private API here...

It's a pity that this is the only substantial problem I have with the example. @amueller, what do you think of making kmeans++ public?

Alternatively, we could consider:
* storing the initial means on the GMM estimator
* allowing GMM to work with  max_iter=0

Reply:
This is a blocker, unfortunately.

Reply:
Do you intend to try fixing this?

Reply:
@amueller what do you think making kmeans++ public?
But I don't mind allowing max_iter=0 either as a diagnostic tool

Reply:
I've added allowing max_iter=0 and changed the plot to not include the private kmeans. Do you think I should add a warning about running it with max_iter=0 or do you think the convergence warning is enough?